consensus/src/chained_bft/event_processor.rs

```diff
@@ -63,6 +65,79 @@ mod event_processor_test;
 #[path = "event_processor_fuzzing.rs"]
 pub mod event_processor_fuzzing;
 
+/// During the event of the node can't recover from local data, StartupSyncProcessor is responsible
+/// for processing the events carrying sync info and use the info to retrieve blocks from peers
+#[allow(dead_code)]
+pub struct StartupSyncProcessor<T> {
+    network: NetworkSender,
+    storage: Arc<dyn PersistentLivenessStorage<T>>,
+    ledger_recovery_data: LedgerRecoveryData<T>,
+}
+
+#[allow(dead_code)]
+impl<T: Payload> StartupSyncProcessor<T> {
+    pub fn new(
+        network: NetworkSender,
+        storage: Arc<dyn PersistentLivenessStorage<T>>,
+        ledger_recovery_data: LedgerRecoveryData<T>,
+    ) -> Self {
+        StartupSyncProcessor {
+            network,
+            storage,
+            ledger_recovery_data,
+        }
+    }
+
+    pub async fn process_proposal_msg(
+        &mut self,
+        proposal_msg: ProposalMsg<T>,
+    ) -> Result<RecoveryData<T>> {
+        let author = proposal_msg.proposer();
+        let sync_info = proposal_msg.sync_info();
+        self.sync_up(&sync_info, author).await
+    }
+
+    pub async fn process_vote(&mut self, vote_msg: VoteMsg) -> Result<RecoveryData<T>> {
+        let author = vote_msg.vote().author();
+        let sync_info = vote_msg.sync_info();
+        self.sync_up(&sync_info, author).await
+    }
+
+    pub async fn process_sync_info_msg(
+        &mut self,
+        sync_info: SyncInfo,
+        peer: Author,
+    ) -> Result<RecoveryData<T>> {
+        debug!("Startup Sync received a sync info msg: {}", sync_info);
+        self.sync_up(&sync_info, peer).await
+    }
+
+    async fn sync_up(&mut self, sync_info: &SyncInfo, peer: Author) -> Result<RecoveryData<T>> {
+        ensure!(
+            sync_info.highest_round() < self.ledger_recovery_data.commit_round(),
+            "Received sync info has lower round number than committed block"
+        );
+        ensure!(
+            sync_info.epoch() == self.ledger_recovery_data.epoch(),
```

Comment:
This part looks suspicious. What happens when you receive a message from some future epoch? I hope we still have a flow for the EpochManager to process different epoch. How is it going to work?

Reply:
Discussed offline, in the startup sync event loop, we will have logic to handle epoch change(via epoch manager), it's essentially the same as the main event processor loop, just we will exist early once we have an event processor constructed by the epoch manager

Reply:
Future epoch messages should be handled outside of EventProcessor layer, today network'd do the dispatch, it's find to move it to EpochManager but that doesn't impact this

Reply:
I imagine it will be something like this:
```
select! {
          startup_sync_processor.process_vote_msg
          startup_sync_processor.process_proposal_msg
          ...
         // Getting a epoch change msg
         ledger_info = network_receivers.epoch_change.select_next_some() => {
                        idle_duration = pre_select_instant.elapsed();
                        if epoch_manager.epoch() <= ledger_info.ledger_info().epoch() {
                            event_processor = epoch_manager.start_new_epoch(ledger_info).await;
                            // clean up all the previous messages from the old epochs
                            network_receivers.clear_prev_epoch_msgs();
                            // exist the loop here
                        }
                    }
          different_epoch_and_peer = network_receivers.different_epoch.select_next_some() => {
                        idle_duration = pre_select_instant.elapsed();
                        epoch_manager.process_different_epoch(different_epoch_and_peer.0, different_epoch_and_peer.1).await
                    }
}
```